### PR TITLE
#413: Load plugins from JAR files

### LINKED ISSUE #413: Load custom plugins from JAR files
OFT already includes some plugins (e.g. importers for RST & MD). These plugins are loaded automatically from the fat JAR that includes all plugins. Using OFT's plugin API, third parties could create and publish and use their own plugins independently from the OFT project. Example: AsciiDoc importer (https://github.com/itsallcode/openfasttrace/issues/410).

## Proposed solution
* OFT loads plugins from folder `OFT_PLUGINS` (e.g. `$HOME/.openfasttrace/plugins/`)
* If the plugin is a single JAR that does not need dependencies, it is directly loaded from `$OFT_PLUGINS/${PLUGIN_NAME}.jar`
* If the plugin requires multiple JARs, OFT loads all JARs from `$OFT_PLUGINS/${PLUGIN_NAME}/*.jar`

## Out of scope
* Configuration of plugin folder via CLI option (maybe in another PR)

## Tasks
- [x] Update loader to load plugins from external JARs
- [ ] Plugin developer guide: document how to write plugins
- [ ] User guide: document how to use plugins
- [ ] User guide: add page with list of third party plugins
- [ ] Readme: add links to list of third party plugins, plugin user guide & plugin developer guide

### PR DESCRIPTION
**Pull request by kaklakariada**
Closes #413

### CODE REVIEW

File: core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceLoaderFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    @TempDir
+    Path tempDir;
+
+    @Test
+    void noPluginJar()
+    {
+        assertThat(testee().createLoader(ReporterFactory.class).load().toList(), empty());
+    }
+
+    private ServiceLoaderFactory testee()
```
Comment (redcatbear) [inline core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceLoaderFactoryTest.java:29]:
```suggestion
    private ServiceLoaderFactory factory()
```

File: core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceLoaderFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        return new ServiceLoaderFactory(tempDir, true);
+    }
+
+    @Test
+    void findServiceSkipCurrentClassLoader() throws IOException
+    {
+        Files.delete(tempDir);
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceLoaderFactoryTest.java:37]: We discussed this and the intention was actually to use a non-existent.

So I would be better to use `resolve("non-exsistent-dir")` in the temp dir.

File: core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceLoaderFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        final List<ServiceOrigin> origins = testee().findServiceOrigins();
+        assertNoPlugins(origins);
+    }
+
+    private void assertNoPlugins(final List<ServiceOrigin> origins) throws MultipleFailuresError
+    {
+        assertAll(() -> assertThat(origins, hasSize(1)),
+                () -> assertThat(origins.get(0).getClassLoader().getName(), equalTo("app")));
```
Comment (redcatbear) [inline core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceLoaderFactoryTest.java:52]: Extract matcher from second assert. Wrap in `contains()`. Remove `hasSize(1)`.

File: core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceLoaderFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Test
+    void findServiceOriginsPluginDir()
```
Comment (redcatbear) [inline core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceLoaderFactoryTest.java:56]: If I look at the code below, this tests behavior on an *empty* plugins parent dir.

File: core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceOriginTest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+class ServiceOriginTest
+{
+    @Test
+    void closesUrlClassLoader(@Mock final URLClassLoader classLoaderMock)
+            throws Exception
+    {
+        final ServiceOrigin serviceOrigin = new ServiceOrigin.JarFileOrigin(Collections.emptyList(), classLoaderMock);
+        serviceOrigin.close();
+        verify(((AutoCloseable) classLoaderMock)).close();
+        verifyNoMoreInteractions(classLoaderMock);
```
Comment (redcatbear) [inline core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceOriginTest.java:28]: :+1: Good. Normally I am not a fan of `verifyNoMoreInteractions`, but here it is absolutely justified.

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceLoaderFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+    @Test
+    void findServiceSkipCurrentClassLoader() throws IOException
+    {
+        Files.delete(tempDir);
+        assertThat(new ServiceLoaderFactory(tempDir, false).findServiceOrigins(), empty());
+    }
+
+    @Test
+    void findServiceMissingParentDir() throws IOException
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply core/src/test/java/org/itsallcode/openfasttrace/core/serviceloader/ServiceLoaderFactoryTest.java]: I renamed the test to `findServiceMissingParentDirFindsNoPlugins()`

**Review by redcatbear — APPROVED** [review]

**Pull request merged**